src/lexer/token.py

```diff
@@ -1,43 +1,119 @@
+from constants.constants import DELIMS
+
+
+class TokenTypes():
+    class TokenType:
+        def __init__(self, token: str, delim_id: str, error_type: str = None):
+            self.token = token
+            self.delim_id = delim_id
+            self.error_type = f"UNTERMINATED {self.token.replace('_', ' ')}" if error_type is None else error_type
+
+            self.expected_delims = DELIMS[self.delim_id]
+
+        def __str__(self):
+            return self.token
```

Comment:
Not sure if others would agree, one suggestion lang to, let's make this a standlone class.

I-move natin yung instances ng `TokenType` sa `constants/token.py`, 
```suggestion
class TokenType():
       def __init__(self, token: str, delim_id: str, error_type: str = None):
            self.token = token
            self.delim_id = delim_id
            self.error_type = f"UNTERMINATED {self.token.replace('_', ' ')}" if error_type is None else error_type

            self.expected_delims = DELIMS[self.delim_id]

       def __str__(self):
            return self.token
```

Reply:
This is what I had in mind:
![image](https://github.com/Gidsss/UwUIDE/assets/70326902/eae45f4d-2887-4357-bbd5-ed5ab305eef4)

Reply:
Imports:
![image](https://github.com/Gidsss/UwUIDE/assets/70326902/ecf89297-148b-46fa-afa1-c445064801b2)

Reply:
Usage of TokenTypes:
![image](https://github.com/Gidsss/UwUIDE/assets/70326902/6a72fc12-0c2f-411a-b2d3-a76da4d06f51)

Usage of TokenType:
![image](https://github.com/Gidsss/UwUIDE/assets/70326902/092996d0-558a-4547-91e7-5933e70d476c)

Reply:
I think that's smart. Constants din kasi ung token types so it makes sense to put them under the constants folder. If we make this change, though, am I right to think na we can just remove the `TokenTypes` superclass? The way we'll fetch them would be like `token_types.<TOKEN_TYPE` where `token_types` is the python file under the constants folder. Or at least, that's how I'm imagining it.

Also, can you make this commit to this branch or do I need to be the one to do it?

Reply:
@HolyShaq Yes, we'll remove `TokenTypes` superclass if we do this change. It would be solely the `TokenType` class nalang wherein we can create instances of it or use it as a type.

When fetching constant values, all you need to do is import the file and give it an alias for readability:
```py
import constants.token as TokenTypes # or other alias names
```

Python is smart enough to group them, I read something about it being related to namespace (idk what that means, maybe cid knows). But it would do the same function you've currently implemented i.e. `TokenTypes.< TOKEN_TYPE >`.

I can commit it pero to avoid excessive commits, you can do it @HolyShaq on your end muna para you can explore if its a better approach or it would cause us problems going further.

Reply:
I think having a token type and token types will be only easy for us since we know it but confusing in a general sense. Separating them won't solve the issue either, just make it so it's getting redirected

I think we should keep the token constants within the class but change the implementation to:

```
from enum import Enum

class TokenType(Enum):
    START = ("START_KEYWORD", "start_done")
    #other tokens ...
    
    def __init__(self, token: str, delim_id: str, error_type: str = None):
        self._token = token
        self._delim_id = delim_id
        self._error_type = f"UNTERMINATED {self.token.replace('_', ' ')}" if error_type is None else error_type
        self._expected_delims = DELIMS[self.delim_id]

    @property
    def token(self):
        return self._token

    # other properties ...

print(f"{TokenType.START.value=}")
print(f"{TokenType.START.token=}")
```

output
```
TokenType.START.value=("START_KEYWORD", "start_done")
TokenType.START.token=START_KEYWORD
```

This way we have less changes with implementation being cleaner

Reply:
I think having a token type and token types will be only easy for us since we know it but confusing in a general sense. Separating them won't solve the issue either, just make it so it's getting redirected

I think we should keep the token constants within the class but change the implementation to:

```
from enum import Enum

class TokenType(Enum):
    START = ("START_KEYWORD", "start_done")
    #other tokens ...
    
    def __init__(self, token: str, delim_id: str, error_type: str = None):
        self._token = token
        self._delim_id = delim_id
        self._error_type = f"UNTERMINATED {self.token.replace('_', ' ')}" if error_type is None else error_type
        self._expected_delims = DELIMS[self.delim_id]

    @property
    def token(self):
        return self._token

    # other properties ...

print(f"{TokenType.START.value=}")
print(f"{TokenType.START.token=}")
```

output
```
TokenType.START.value=("START_KEYWORD", "start_done")
TokenType.START.token=START_KEYWORD
```

This way we have less changes with implementation being cleaner. Please see [this](https://docs.python.org/3/howto/enum.html#planet) for sample of this exact usage of enum in python docs

Reply:
I'm g sa sinabi ni Cid. I think yung paggamit ng "Enum" for token types provides a more clean and structured way to define and represent a fixed set of token types(?). This is just my opinion tho.

Reply:
@am-cid thanks, I learned something new! I think this is a better approach @HolyShaq . 

I tested it and does the similar idea I had in mind earlier before making this suggestion. I wanted the `TokenTypes` class to be able to create instances of itself within the class using `@classmethod` pero it did not work:
```py
# What I had in mind earlier
class TokenTypes():
    def __init__(self, token: str, delim_id: str, error_type: str = None):
        self._token = token
        self._delim_id = delim_id
        self._error_type = f"UNTERMINATED {self.token.replace('_', ' ')}" if error_type is None else error_type
        self._expected_delims = DELIMS[self.delim_id]
    
    @classmethod
    def TokenType(cls, *args):
         return cls(*args)

    START = self.TokenType("START_KEYWORD", "start_done")
```

I like Cid's implementation for this and I agree na its cleaner

Reply:
Thanks y'all. I didn't know that you can instantiate a class within itself.
This is very clean!! `START = ("START_KEYWORD", "start_done")`
I'm happy to get rid of my subclass any day of the week HAHAHAHAHAH

So in the end, we keep the `TokenType` class under `token.py`. Then just change how the `TokenType` constants are stored neh? Will work on it na. Also thought of a possible refactor with the error type habang naliligo. Pacheck nalang once I've commited it!

Reply:
Btw, @am-cid is there any particular reason why self.token should be a property?

Reply:
no reason really. it'll work fine either way.

you can use self.token as an instance attribute within `__init__` but I wanted to make it explicit that you cannot set values to it (since I'm not putting `@token.setter` only a `@property` which is a getter by default)

but anw, even if you use self.token as an instance attribute within `__init__` and try to set values to it, python will yell at you either way and say that you cannot assign values to an enum! doing it the `@property` way will make the python yell something else, I forgot